ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/HD/DbChangelog.hs

```diff
@@ -0,0 +1,430 @@
+-- This means that after a rollback of @k@ blocks at (*), the changelog will
+-- look something like this:
+--
+-- >      2      |      0     | [ L0, L1, L2 ]             | [ D1, D2 ]
+--
+-- And a rollback of @k@ blocks at (**) will look something like this:
+--
+-- >      2      |      0     | [ L2 ]                     | [ ]
+--
+-- Notice how the states list always contains the in-memory state of the anchor,
+-- but the table differences might not contain the differences for that anchor
+-- if they have been flushed to the backend.
+--
+-- As said above, this @DbChangelog@ has to be coupled with a @BackingStore@
+-- which provides the pointers to the on-disk data.
+--
+-- INVARIANT: the head of 'changelogImmutableStates' is the anchor of
+-- 'changelogVolatileStates'.
+data DbChangelog l = DbChangelog {
+    changelogDiffAnchor      :: !(WithOrigin SlotNo)
+  , changelogDiffs           :: !(LedgerTables l SeqDiffMK)
+  , changelogImmutableStates ::
+      !(AnchoredSeq
+          (WithOrigin SlotNo)
+          (DbChangelogState l)
+          (DbChangelogState l)
+       )
+  , changelogVolatileStates  ::
+      !(AnchoredSeq
+          (WithOrigin SlotNo)
+          (DbChangelogState l)
+          (DbChangelogState l)
+       )
+  }
+  deriving (Generic)
+
+deriving instance (Eq       (LedgerTables l SeqDiffMK), Eq       (l EmptyMK))
+               =>  Eq       (DbChangelog l)
+deriving instance (NoThunks (LedgerTables l SeqDiffMK), NoThunks (l EmptyMK))
+               =>  NoThunks (DbChangelog l)
+deriving instance (Show     (LedgerTables l SeqDiffMK), Show     (l EmptyMK))
+               =>  Show     (DbChangelog l)
+
+newtype DbChangelogState l = DbChangelogState {unDbChangelogState :: l EmptyMK}
+  deriving (Generic)
+
+deriving instance Eq       (l EmptyMK) => Eq       (DbChangelogState l)
+deriving instance NoThunks (l EmptyMK) => NoThunks (DbChangelogState l)
+deriving instance Show     (l EmptyMK) => Show     (DbChangelogState l)
+
+instance GetTip (l EmptyMK) => AS.Anchorable (WithOrigin SlotNo) (DbChangelogState l) (DbChangelogState l) where
+  asAnchor = id
+  getAnchorMeasure _ = getTipSlot . unDbChangelogState
+
+{-------------------------------------------------------------------------------
+  Construction
+-------------------------------------------------------------------------------}
+
+empty ::
+     (HasLedgerTables l, GetTip (l EmptyMK))
+  => l EmptyMK -> DbChangelog l
+empty anchor =
+    DbChangelog {
+        changelogDiffAnchor      = getTipSlot anchor
+      , changelogDiffs           = pureLedgerTables (SeqDiffMK DS.empty)
+      , changelogImmutableStates = AS.Empty (DbChangelogState anchor)
+      , changelogVolatileStates  = AS.Empty (DbChangelogState anchor)
+      }
+
+{-------------------------------------------------------------------------------
+  Updates
+-------------------------------------------------------------------------------}
+
+extend ::
```

Comment:
Here we could benefit from some examples. Maybe we should not do them in this PR, but we could probably add them to a larger issue that keeps track of the efforts for documenting UTxO HD.

Reply:
Addressed in #4392